game_template/lib/src/crashlytics/crashlytics.dart

```diff
@@ -0,0 +1,96 @@
+///
+/// These errors will also include latest logs from anywhere in the app
+/// that use `package:logging`.
+Future<void> guardWithCrashlytics(
+  void Function() mainFunction, {
+  required FirebaseCrashlytics? crashlytics,
+}) async {
+  // Running the initialization code and [mainFunction] inside a guarded
+  // zone, so that all errors (even those occurring in callbacks) are
+  // caught and can be sent to Crashlytics.
+  await runZonedGuarded<Future<void>>(() async {
+    if (kDebugMode) {
+      // Log more when in debug mode.
+      Logger.root.level = Level.FINE;
+    }
+    // Subscribe to log messages.
+    Logger.root.onRecord.listen((record) {
+      final message = '${record.level.name}: ${record.time}: '
+          '${record.loggerName}: '
+          '${record.message}';
+
+      debugPrint(message);
+      // Add the message to the rotating Crashlytics log.
+      crashlytics?.log(message);
+
+      if (record.level >= Level.SEVERE) {
+        crashlytics?.recordError(message, filterStackTrace(StackTrace.current));
+      }
+    });
+
+    // Pass all uncaught errors from the framework to Crashlytics.
+    if (crashlytics != null) {
+      WidgetsFlutterBinding.ensureInitialized();
+      FlutterError.onError = crashlytics.recordFlutterError;
+    }
+
+    // To catch errors outside of the Flutter context, we attach an error
+    // listener to the current isolate.
+    Isolate.current.addErrorListener(RawReceivePort((dynamic pair) async {
+      final errorAndStacktrace = pair as List<dynamic>;
+      await crashlytics?.recordError(
+        errorAndStacktrace.first,
+        errorAndStacktrace.last as StackTrace?,
+      );
+    }).sendPort);
+
+    // Run the actual code.
+    mainFunction();
+  }, (error, stack) {
+    // This sees all errors that occur in the runZonedGuarded zone.
+    debugPrint('ERROR: $error\n\n'
+        'STACK:$stack');
+    crashlytics?.recordError(error, stack);
+  });
+}
+
+/// Takes a [stackTrace] and creates a new one, but without the lines that
+/// have to do with this file and logging. This way, Crashlytics won't group
+/// all messages that come from this file into one big heap just because
+/// the head of the StackTrace is identical.
+///
+/// See this:
+/// https://stackoverflow.com/questions/47654410/how-to-effectively-group-non-fatal-exceptions-in-crashlytics-fabrics.
+@visibleForTesting
+StackTrace filterStackTrace(StackTrace stackTrace) {
+  StackTrace? filtered;
+  try {
+    final lines = stackTrace.toString().split('\n');
+    final buf = StringBuffer();
+    for (final line in lines) {
+      if (line.contains('crashlytics.dart') ||
+          line.contains('_BroadcastStreamController.java') ||
+          line.contains('logger.dart')) {
+        continue;
+      }
+      buf.writeln(line);
+    }
+    filtered = StackTrace.fromString(buf.toString());
+  } catch (e) {
+    debugPrint('Problem while filtering stack trace: $e');
+  }
+
+  return filtered ?? stackTrace;
```

Comment:
Can just skip the variable:

```suggestion
  try {
    final lines = stackTrace.toString().split('\n');
    final buf = StringBuffer();
    for (final line in lines) {
      if (line.contains('crashlytics.dart') ||
          line.contains('_BroadcastStreamController.java') ||
          line.contains('logger.dart')) {
        continue;
      }
      buf.writeln(line);
    }
    return StackTrace.fromString(buf.toString());
  } catch (e) {
    debugPrint('Problem while filtering stack trace: $e');
  }
  
  // If there was an error while filtering,
  // return the original, unfiltered stack track.
  return stackTrace;
```

Reply:
Thanks, you opened my mind here. All this time I was avoiding control statements anywhere in `try`/`catch`, but it turns out it only makes sense to avoid them in `finally`, and it's completely fine to have them in `try`.